text/0000-signals-in-lwc-templates.md

```diff
@@ -0,0 +1,196 @@
+---
+title: Supporting Signals in LWC Templates
+status: DRAFT
+created_at: 2023-11-15
+updated_at: 2023-11-15
+champion: Caridy & Dale 
+pr: https://github.com/salesforce/lwc-rfcs/pull/
+---
+
+# Supporting Signals in LWC Templates
+
+## Summary
+
+This RFC proposes an enhancement to the rendering mechanism in Lightning Web Components (LWC) to support Signals-based objects in LWC Templates, this is inspired by similar mechanisms in libraries such as Preact, vue, etc. The goal is to provide a flexible mechanism to integrate stores or signals into the template in such a way that the rehydration of the UI is still possible even when the user is not relying on the observable membrane provided by LWC. The introduction of Signals in LWC opens the door for more ergonomic methods of observing state changes in the component and allows developers to bring their own data stores if they adhere to a simple protocol involving `.value`, and `.subscribe` properties. This change can also accelerate the work on performance optimizations within LWC Rendering process, especially when rehydrating the template, potentially reducing the need for full component re-rendering through a more efficient subscription mechanism.
+
+## Motivation
+
+The current reactivity model in LWC relies solely in the observable membrane, via proxies, which are tightly coupled with LWC internals. This coupling leads to certain limitations:
+
+1. __Limited External Store Interaction__: LWC components cannot directly interact with external state management systems without going through LWC-specific APIs. This restricts the framework's flexibility and developer freedom to integrate with the broader JavaScript ecosystem.
+1. __Complexity and Overhead__: The use of a membrane based on proxies introduces additional complexity and overhead in component design and state management, from read-and-write to read-only objects, it can be challenging for developers to debug and reason about data flow and updates.
```

Comment:
This is mostly about cognitive overhead for developers, not performance overhead at runtime, yes?

I assume the main issue here is how DevTools represents Proxies, having to turn on DevTools custom formatters, etc.? Or is this also about tracking how changes flow through the reactivity system? (I.e. the fact that we're missing something akin to Redux DevTools?)

Reply:
@nolanlawson No, the main point here is that if you choose to use `@track`, it affects your design. Now, if the child component is expecting the object, but it should have influence over the values on that object, then you need to create a channel in user-land (probably an event), so the owner of the object has to modify the read-write version of the same object. And of course, this makes it harder to debug how the data flows.